### ISSUE TITLE
Add MP3 support

### ISSUE THREAD

**Issue by erikd** [libsndfile/libsndfile#258]
The MP3 patents have now fully expired so libsndfile can now add support. 

The best bet would be https://www.mpg123.de which is LGPL and has updates as recently as this year.

I (@erikd) don't have sufficient time and motivation to do this, but would be happy to guide someone else.

**Comment by evpobr** [comment]
Milestone?

**Comment by erikd** [comment]
Until some actually steps forward to offer to do this and actually has some code that is worth reviewing, it doesn't make sense to have a milestone.

**Comment by dabfalcon** [comment]
I hope someone steps forward and this feature would be great.

**Comment by erikd** [comment]
It also seems that https://www.mpg123.de is a decoder only. All the other supported formats have both encode and decode so that would have to be the same for MP3 especially fore the CSound use case.

liblame may be the best bet.

**Comment by evpobr** [comment]
Agreed, `mpg123` and `LAME` are the best.

Bad thing i will be buzy all the summer.

**Comment by dabfalcon** [comment]
Bad thing I'm too n00b. How hard is it?
I'm good with PHP / web / did a course on C++.

**Comment by erikd** [comment]
This is definitely more difficult than using libsndfile as a library, mainly because the libsndfile API was specifically designed to be easy to use. I would also rank this as more difficult than adding Opus support, because it requires you to hook up two libraries, one for decoding and one for encoding.

**Comment by faroit** [comment]
Also interesting, from [this recent statement](http://www.audioblog.iis.fraunhofer.com/mp3-software-patents-licenses/):

> However, the end of the mp3 licensing program does not automatically mean that all mp3 technology is available license-free now. Apart from the core mp3 patents included in the licensing program, there might still be some implementation–specific patents (or patents for other functional enhancements) that have not expired. Thus, manufacturers will have to check the situation regarding their intended products first before including mp3.

**Comment by jpffitch** [comment]
In csound at present we use a Russian library libmpadec for both read ad write.  Seems OK.Is it worth investigating?  I should add that it  has not been changed by the author for years but i have tweaked it, and have developer status for it.  Code in csound tree and on sourceforge

**Comment by jpffitch** [comment]
Is there a template for adding a new format or does one just edit, say, ogg.c to new code?  Some time since I last tried (10 years) and things have changed.

**Comment by erikd** [comment]
Closest thing to a template is `src/new.c`.

**Comment by eblanca** [comment]
For a wider view, i see on the decoding side there's much more: there is libmad which is very VERY stable and currently many distributions still provide it, not to mention its feature being based only on integer number calculation. This makes it very suitable for embedded systems, imho.
On the other hand, the writing/encoding side also offers many names to pick. If libsndfile will be able to read mpeg layer1/2/3 audio, then why offering layer3-only encoding? libtwolame provides layer2 encoding (layer2 is very often used even today, both into DVDs and into DVB streams) and for layer1, well, perhaps it may be also evaluated.

**Comment by jpffitch** [comment]
Does the licence matter?  GPL v LGPL ?  With so many options I feel the need for guidance.

The mpg123 +lame plan looks like the mainsteam way to go, but I do not claim expertise in MP3.

**Comment by erikd** [comment]
GPL is no good, but LGPL, BSD, MIT would all be fine.

**Comment by rjamorim** [comment]
> not to mention its feature being based only on integer number calculation. This makes it very suitable for embedded systems, imho.

mpg123 can be built to perform only fixed-point math, and with better performance than MAD according to Thomas Orgis. Also, mpg123 has assembly optimizations for several architectures and SIMD flavors.

**Comment by eblanca** [comment]
> mpg123 can be built to perform only fixed-point math

Didn't know this, great!

**Comment by cvanwinkle** [comment]
FWIW, I've used mpg123 under LGPL in my program for MPEG-2 Audio Layer II decoding (aka MP2 audio) for a while and have validated that it decodes MP3 to my liking although we don't use it for that. So one nice thing about mpg123 is that it can decode several layers of MPEG audio and not just what is known as MP3. So assuming someone else does the work for wiring in mpg123 other than me (which is likely), it would be good to accommodate the other flavors as well.

**Comment by rjamorim** [comment]
Hello, Erik. Indeed, mpg123 only does decoding, and I strongly support using LAME for encoding. Back in 2004 I conduced a public listening test, and LAME got the best ranking (granted, it was a long time ago, but neither of the other encoders advanced much since then)
https://web.archive.org/web/20071016053926/http://www.rjamorim.com:80/test/mp3-128/results.html

Also, another reason I thought for choosing mpg123 over MAD (besides the licensing issue): MAD was last updated in 2004. mpg123 was last updated last week and the developer, Thomas Orgis, is very approachable.

**Comment by foolswood** [comment]
I am happy to have a go at this (unless someone else has started in which case let me know).

**Comment by foolswood** [comment]
Made a start (not working yet) over here: https://github.com/foolswood/libsndfile/tree/mp3.

Thinking about RDWR, that might be a bit tricky with the lame/mpg123 split, any thoughts on that (otherwise I might leave that mode unimplemented in the first attempt).

**Comment by foolswood** [comment]
Read support PR is #322 at the moment I'm after shape review to see if I'm on the right track but hopefully it makes at least some sense :)

**Comment by foolswood** [comment]
Lame can use libsndfile for opening its input formats, so I'm concerned that using it as a decoder for libsndfile introduces a dependency loop. If lame were to split into liblame and the lame frontend (as I believe it is the latter that has the dependency) that would break the cycle. However given that's a big change and the last release was in 2012 I don't think that's likely.

Suggestions welcome.

**Comment by foolswood** [comment]
It has been suggested that including a copy of lame.h with libsndfile and loading at runtime with dlopen (and erroring if not found) makes it a runtime dependency and this may be a way to break the cycle.

@erikd Does this seem like an acceptable solution?

**Comment by erikd** [comment]
> Lame can use libsndfile for opening its input formats

The important part there is *can*. Ie its optional.

> I'm concerned that using it as a decoder for libsndfile introduces a dependency loop.

Since is optional, there is no dependency loop.

> last release was in 2012

I am *really* wary of depending a a piece of software that has not be updated since 2012.

**Comment by rjamorim** [comment]
> I am really wary of depending a a piece of software that has not be updated since 2012.

Lame is being updated again, release 3.100 should happen soon:
http://lame.cvs.sourceforge.net/viewvc/lame/lame/doc/html/history.html
(you can also check the mailing list to see the activity going on)

And either way, if you want to add open source MP3 encoding, there's little choice other than LAME - actively maintained or not.

**Comment by lieff** [comment]
There also libshine encoder https://github.com/toots/shine
It's included in modern linux distributions and still maintained.

**Comment by ousia** [comment]
@lieff, although shine seems to be [much faster](https://github.com/toots/shine#how-fast-is-it) than Lame, code is [rather old](https://github.com/toots/shine#a-bit-of-history) and [limited](https://github.com/toots/shine#limitations).

Lacking a psychoacoustic model may affect to file compression or sound quality (but I’m not an expert).

**Comment by ousia** [comment]
@erikd, I have just discovered that `lame-3.100` was released in October 2017 (http://lame.sourceforge.net/).

**Comment by kcat** [comment]
As an alternative to libmpg123 for decoding, there's [minimp3](https://github.com/lieff/minimp3/). It's a much smaller code base (a single header, no external libraries), doesn't worry about a bunch of old micro-optimized assembly functions for 3/4/586 CPUs (that a modern compiler optimizer could probably outdo anyway), and performance is perfectly fine (utilizes SSE and NEON intrinsics when available). It's public domain (or CC0) so licensing isn't an issue. There's also currently a PR to get the "pure" floating-point samples generated from it, without clamping or 16-bit quantization noise from a `float`->`short`->`float` conversion.

**Comment by lminer** [comment]
Any updates on this? Is it dead?

**Comment by erikd** [comment]
I am not aware of anyone active working on this at the moment.

**Comment by jpffitch** [comment]
It might be useful for a recapitulation of the last discussion.  I seem to 
have mislaid the e-mails but the issues seemed to be that there was no 
simle library fotr both in and out of MP3 files and so two libraries were 
needed.  What I cannot recall is which libraries seemed the best for the 
job.

I am interested in this project, not least to simplify csound.  I may be 
rather slow in coding at present but I would like to encourage the work.

==John ff

**Comment by rjamorim** [comment]
> What I cannot recall is which libraries seemed the best for the job. I am interested in this project, not least to simplify csound.

LAME is the only real choice for MP3 encoding, really. Shine is too limited in features (also, quality is really bad since there is no psychoacoustic processing whatsoever). You can ask Gabriel Bouvigne, he was a LAME developer and the creator of Shine.

As for MP3 decoding, it seems the choices boil down to libmpg123 (time tested, used everywhere, code can be considered bloated with all those ASM optimizations) and minimp3 (I can't comment much)

**Comment by jpffitch** [comment]
I did not manage to find a definition of an API for LAME.  What have I missed?

**Comment by njh** [comment]
@jpffitch there is some documentation for the LAME API here:
https://sourceforge.net/p/lame/svn/HEAD/tree/trunk/lame/API

And the full API is described in the header file here:
https://sourceforge.net/p/lame/svn/HEAD/tree/trunk/lame/include/lame.h

There is an MPEG Audio decoder in the LAME library:
"Based off of Michael Hipp's [mpglib](http://www.mpg123.de/), with many
improvements by the lame development team"
https://sourceforge.net/p/lame/svn/HEAD/tree/trunk/lame/mpglib/

@erikd Does there need to be a single dependency for both encoding and decoding MP3s?
Can we start with a PR for either encoding/decoding, or does it have to support both from the start?


I would also quite like to have MPEG Audio Layer 2 encoding support in libsndfile. For my project [rotter](https://github.com/njh/rotter/), I created a wrapper, that had support for writing files to MP2, MP3 and some of the formats supported by libsndfile - it would have been much easier if libsndfile supported MPEG Audio.

**Comment by szabyg** [comment]
[GSOC](https://summerofcode.withgoogle.com/) could be an idea if time is the issue. Would need someone to mentor.

**Comment by jpffitch** [comment]
I looked at minimp3 and it looked fairly straightforward.  OTOH we in csound have been using a library called libmpadec for decoding  After a few minor tweaks it is working quite well so I might look at that route.

Currently trying to clear projects of my virtual desk to look at the LAME API -- thanks for the link.

**Comment by jpffitch** [comment]
I have been looking at the code and realise that it is12 years since I wrote code here and the build system has changed.  I have not yet managed to add a new file to the system to handle a new format.  Is there any instructions/notes/documentation or do I just hack around until it compiles?

**Comment by jpffitch** [comment]
I do not seem to be able to mail the development list so..  I have code written to open a MP3 file for output but am having trouble witha few things.  

How can i compile with -g option so gdb can single step?

What are the psf fields datalength and dataoffset for?  I set them to 0 so I could avoid checks after the new code but now I get SFE_UNIMPLEMENTED and not sure from whence it comes.

MP3 has three controls, quality, bitrate and mode (mono stereo,joint) .  I am not sure how to specify these

**Comment by jpffitch** [comment]
I can create MP3 files in various formats and write audio data apparently working.  Not sure what to try next.

Sil need to solve the quality and brtrate parameters.

**Comment by jpffitch** [comment]
As far as I have been able to find out looking at flac/ogg/... the policy is to use sf_command to set quality and bitrate. So I need two new commands to implement that, or to reuse commands SFC_SET_VBR_ENCODING_QUALITY and SFC_SET_COMPRESSION_LEVEL.  The first look rather specific but the latter would encode the quality value (which is not really compression level though....)

I have been wondering about decode.  Choice seems to be minimp3, mpg123 or the LAME decoder.  The first looks feasible as far as I have got, the second may already be done by foolswood.  I know Erik had issues with the pull request but perhaps that could be used?  (I will try to d.load it.) The LAME route had the least documentation/examples.  I am seeking advice as which way to go.

**Comment by jpffitch** [comment]
Tidied up the code, fixed a distortion problem and wrote commands to control quality and bit rate.
Until others try it I am freezing the output code.

Not managed to work out how to make it available.  Trying to understand pull requests but I seem as ever to fail to follow documention. 

files changed (minor) are
CMakeLists.txt, src/common.h,  src/sndfile.c and sndfile.h.in
Two new files, src/mp3.c and src/mp3.h
Requires libmp2lame and lame.h

**Comment by jpffitch** [comment]
Just a small update.  I can read some MP3 files with sf_read_short but there is a silence at te start of about 1100 samples and the stereo version seems suspect.  Writing of MP3 seems OK now.  I have a few questions I will pose as soon as I am a little further.  My day has ended.

The is using LAME (libmp3lame) for both reading and writing

**Comment by jpffitch** [comment]
Progress.  I seem to be able to read  MP3 files now.  Some more resting and a couple of oddities to iron out before another PR.

**Comment by eblanca** [comment]
@jpffitch what about renaming your new files to "mpega.c" and "mpega.h" rather than "mp3.*" ?

After all, the reading library (whatever you pick) supports any mpeg layer, does it ?

**Comment by jpffitch** [comment]
Current state, using mpg123 for reading and lame for writing basic read and write work.  There are a few loose ends but it is in reasonable state.  I would like to thank @foolswood for his work on mpg123 which has saved me a lot of time.
Time to test and tidy.

**Comment by martinwguy** [comment]
Many thanks to John and foolswood for tackling this (finally!)

I've tried compiling and it needed some build system tweaks:
In mp3.h, I meeded to include <lame.h> and <mpg123.h>, not "lame.h"
and "mpg123.h", but (on Debian) lame.h is /usr/include/lame/lame.h so,
being ignorant in configure.ac hacking, I changed
```
- MP3_CFLAGS="$MPG123_CFLAGS"
+ MP3_CFLAGS="$MPG123_CFLAGS -I/usr/include/lame"
```
however, the MP3_* defines don't make it into CFLAGS and LDFLAGS, so
another nasty hack:
```
+CFLAGS="$CFLAGS $MP3_CFLAGS"
+LDFLAGS="$LDFLAGS $MP3_LIBS"
```
At this point it compiled (with 5 compiler warnings for mp3.c)

I also noticed that it needs lame 3.100 (which is the version in current Debian stable), as the old Debian I was using had 3.99, which lacks the needed lame_encode_buffer_interleaved_int()

Unfortunately I can't test the MP3 decoding for in my application because seeking in
MP3 files is not implemented, and it needs that. I'm working on this to be able to test it in its current state. Incidentally, seeking in MP3 files is nasty and produces noisy output immediately after the seek, I think because to decode a block of the later MP3 variants, the result depends on a small amount of extra data that is stuffed into a little bit of free space in the previous two blocks.

If a configure wizard could step in, implement --enable-mp3, or possibly --enable-mp3-encoder and enable-mp3-decoder (or all three) and if so, checking that libmp3lame is preset and >=3.100, that would be fab, libmp3lame doesn't seem to have pkg-config magic (libmp3lame.pc), at least at present in Debian.

i don't think including lame.h in libsndfile is a good idea, as it may be (or become) different from the version of lame that is installed on the system.

**Comment by leimao** [comment]
So the current status is "libsndfile still does not support mp3"?

**Comment by erikd** [comment]
Correct. It is being worked on, but it is not complete.

**Comment by leimao** [comment]
> Correct. It is being worked on, but it is not complete.

Thanks a lot. I was looking at some code examples of decoding sound files using C/C++. Looks like the code example using libsndfile is much shorter than the code examples using other libraries such as ffmpeg. Since ffmpeg supported mp3 already, it would be good to see libsndfile also supports mp3 :)

**Comment by erikd** [comment]
> Looks like the code example using libsndfile is much shorter than the code examples using other libraries such as ffmpeg.

libsndfile was designed to have an API that was as easy to use as possible.

> it would be good to see libsndfile also supports mp3

It is being worked (although not by me).

**Comment by nickvaras** [comment]
Kudos for this work guys, sending you all the Chi. Just know that you will make some people quite happy with mp3 support.  :smiley:

**Comment by define-private-public** [comment]
I'd also like to voice my thanks for working on this.  I'm created (as far as I know so far), the only cross platform audio loading and playback library for C#/.NET core and I'm using `libsndfile` right now for all of the media format decoding.  I was looking to use MP3Sharp to do MP3 decoding.  But if I can get MP3 decoding w/ out having to add an extra project, that is super.

**Comment by Jasparr77** [comment]
Hey guys, just checking in on this. How's it going? Any tricky blockers or tasks you want to off-load? Echoing the many thanks you've received on your work already.

**Comment by jpffitch** [comment]
Basically it works.  The only problems of which I am aware is that if 
there is an error (sorry cannot remember which it is -- reading or 
writing) the library prints a message which ideally should be captured by 
libsndfile and treated like all other errors in the system.

That said it could probably do with more use for testing.

**Comment by marcusmueller** [comment]
@jpffitch is your current progress visible somewhere? That would enable me to test, and possibly, improve.

**Comment by martinwguy** [comment]
Thanks! There would also be to modify doc/{index.html,FAQ} to say that
MP3 is supported. Do we know if it supports seeking and is
sample-accurate when doing so?

   M

**Comment by csukuangfj** [comment]
Is there any plan to merge https://github.com/arthurt/libsndfile.git ?

**Comment by AngledLuffa** [comment]
Any progress on this?  Could really use this for a personal project of mine.

**Comment by martinwguy** [comment]
Erik seems not to have much time for libsndfile. My 2015 work giving sndfile-spectrogram a logarithmic frequency axis was committed in 2016 but not released until 2019, and will have to wait many more years before it gets into the mainline distributions. Similarly, the MP3 work was ready in February of this year, but has not yet been merged, even though erik *is* actively merging other smaller changes into the git master, so who knows when it will arrive in a release and, years later, get into the mainline distros?

Something as crucial as MP3 support should be a major priority for a sound file reading library, but its official maintainer, who is also its Debian maintaner, is just too busy with other things, and doesn't seem able to appoint someone with more time as the maintainer.

For your own project, I suggest you just install arthurt's codec-mp3 branch and use that. If you subsequently decide to publish your work, the best strategy woud be to include a custom copy of libsndfile in your source tree and compile against that. Alternatively, use one of the existing decoding libraries for MP3 files, of which libmpg123 has the unique advantage of being able to seek in MP3 files to the accuracy of one sample.

However, for mainline libsndfile, don't hold your breath.

**Comment by evpobr** [comment]
https://github.com/libsndfile/libsndfile/pull/499#issuecomment-700242176

**Comment by randomMesh** [comment]
Looking forward to this. Thumbs up!

**Comment by Nelson-numerical-software** [comment]
Could we consider that this issue is closed with 1.1.0  ? :)

**Comment by evpobr** [comment]
I guess so :smile:

**Comment by AngledLuffa** [comment]
Thank you everyone!  I'm very happy to be able to use this once my work gives me a moment's break

**Comment by martinwguy** [comment]
> Closed #258.

Finally! :) Lots of love to all who helped

    M
